File: app/src/main/java/com/metinkale/prayerapp/vakit/WidgetProviderClock.java
package com.metinkale.prayerapp.vakit;

import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Paint.Align;
import android.graphics.Paint.Style;
import android.os.Bundle;
import android.provider.AlarmClock;
import android.util.TypedValue;
import android.widget.RemoteViews;
import com.crashlytics.android.Crashlytics;
import com.metinkale.prayer.R;
import com.metinkale.prayerapp.HicriDate;
import com.metinkale.prayerapp.Utils;
import com.metinkale.prayerapp.settings.Prefs;
import com.metinkale.prayerapp.vakit.times.Times;
import com.metinkale.prayerapp.vakit.times.Vakit;
import org.joda.time.LocalDate;
import org.joda.time.LocalTime;

public class WidgetProviderClock extends AppWidgetProvider {
    private static float mDP;

    public static void updateAppWidget(Context context, AppWidgetManager appWidgetManager, int widgetId) {

        if (mDP == 0) {
            Resources r = context.getResources();
            mDP = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 1, r.getDisplayMetrics());
        }

        Resources r = context.getResources();
        SharedPreferences widgets = context.getSharedPreferences("widgets", 0);


        int w = widgets.getInt(widgetId + "_width", 500);
        int h = widgets.getInt(widgetId + "_height", 200);

        w = (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, w, r.getDisplayMetrics());
        h = (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, h, r.getDisplayMetrics());

        float scaleX = (float) w / (float) 5;
        float scaleY = (float) h / (float) 2;
        float scale = Math.min(scaleY, scaleX);


        w = (int) (5 * scale);
        h = (int) (2 * scale);

        if (w <= 0 || h <= 0) {
            SharedPreferences.Editor edit = widgets.edit();
            edit.remove(widgetId + "_width");
            edit.remove(widgetId + "_height");
            edit.apply();
            updateAppWidget(context, appWidgetManager, widgetId);
            return;
        }

        Times times = null;
        long id = widgets.getLong(widgetId + "", 0L);
        if (id != 0)
            times = Times.getTimes(widgets.getLong(widgetId + "", 0L));
        if (times == null) {
            RemoteViews remoteViews = new RemoteViews(context.getPackageName(), R.layout.widget_city_removed);
            Intent i = new Intent(context, WidgetConfigureClock.class);
            i.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, widgetId);
            remoteViews.setOnClickPendingIntent(R.id.image, PendingIntent.getActivity(context, 0, i, PendingIntent.FLAG_CANCEL_CURRENT));
            appWidgetManager.updateAppWidget(widgetId, remoteViews);
            return;
        }


        RemoteViews remoteViews = new RemoteViews(context.getPackageName(), R.layout.vakit_widget_clock);

        remoteViews.setOnClickPendingIntent(R.id.abovePart, PendingIntent.getActivity(context, (int) System.currentTimeMillis(), new Intent(AlarmClock.ACTION_SHOW_ALARMS), PendingIntent.FLAG_UPDATE_CURRENT));
        remoteViews.setOnClickPendingIntent(R.id.belowPart, Main.getPendingIntent(times));


        int next = times.getNext();
        int last = next - 1;

        Bitmap bmp = Bitmap.createBitmap(w, h, Bitmap.Config.ARGB_4444);
        Canvas canvas = new Canvas(bmp);
        Paint paint = new Paint();
        paint.setAntiAlias(true);
        paint.setDither(true);
        paint.setFilterBitmap(true);


        paint.setStyle(Style.FILL_AND_STROKE);
        paint.setAntiAlias(true);
        paint.setSubpixelText(true);
        paint.setShadowLayer(2, 2, 2, 0xFF555555);
        paint.setTextAlign(Align.CENTER);
        paint.setColor(Color.WHITE);

        LocalTime ltime = LocalTime.now();

        paint.setTextSize(h * 0.55f);
        String time = ltime.toString("HH:mm");
        if (Prefs.use12H()) {
            time = Utils.fixTime(time);
            String suffix = time.substring(time.indexOf(" ") + 1);
            time = time.substring(0, time.indexOf(" "));
            canvas.drawText(time, w / 2 - paint.measureText(suffix) / 4, h * 0.4f, paint);
            paint.setTextSize(h * 0.275f);
            canvas.drawText(suffix, w / 2 + paint.measureText(time), h * 0.2f, paint);
        } else {
            canvas.drawText(time, w / 2, h * 0.4f, paint);
        }


        LocalDate date = LocalDate.now();
        String greg = Utils.format(date);
        String hicri = Utils.format(new HicriDate(date));

        paint.setTextSize(h * 0.12f);
        float m = paint.measureText(greg + "  " + hicri);
        if (m > w * 0.8f) {
            paint.setTextSize(h * 0.12f * w * 0.8f / m);
        }


        paint.setTextAlign(Align.LEFT);
        canvas.drawText(greg, w * .1f, h * 0.55f, paint);
        paint.setTextAlign(Align.RIGHT);
        canvas.drawText(hicri, w * .9f, h * 0.55f, paint);
        remoteViews.setImageViewBitmap(R.id.widget, bmp);

        canvas.drawRect(w * 0.1f, h * 0.6f, w * 0.9f, h * 0.63f, paint);

        if (times.isKerahat()) {
            paint.setColor(0xffbf3f5b);
        } else {
            paint.setColor(Theme.Light.strokecolor);
        }
        long mills1 = times.getMills(last);
        long mills2 = times.getMills(next);
        long passed = System.currentTimeMillis() - mills1;
        float percent = passed / (float) (mills2 - mills1);
        canvas.drawRect(w * 0.1f, h * 0.6f, w * 0.1f + w * 0.8f * percent, h * 0.63f, paint);

        paint.setColor(Color.WHITE);
        paint.setTextSize(h * 0.2f);
        paint.setTextAlign(Align.LEFT);
        if (Prefs.use12H()) {
            String l = Utils.fixTime(times.getTime(last));
            String s = l.substring(l.indexOf(" ") + 1);
            l = l.substring(0, l.indexOf(" "));
            canvas.drawText(times.getTime(last), w * 0.1f, h * 0.82f, paint);
            paint.setTextSize(h * 0.1f);
            canvas.drawText(s, w * 0.1f + 2 * paint.measureText(l), h * 0.72f, paint);

        } else {
            canvas.drawText(times.getTime(last), w * 0.1f, h * 0.82f, paint);

        }
        paint.setTextSize(h * 0.12f);
        canvas.drawText(Vakit.getByIndex(last).getString(), w * 0.1f, h * 0.95f, paint);

        paint.setColor(Color.WHITE);
        paint.setTextSize(h * 0.2f);
        paint.setTextAlign(Align.RIGHT);
        if (Prefs.use12H()) {
            String l = Utils.fixTime(times.getTime(next));
            String s = l.substring(l.indexOf(" ") + 1);
            l = l.substring(0, l.indexOf(" "));
            canvas.drawText(l, w * 0.9f - paint.measureText(s) / 2, h * 0.82f, paint);
            paint.setTextSize(h * 0.1f);
            canvas.drawText(s, w * 0.9f, h * 0.72f, paint);

        } else {
            canvas.drawText(times.getTime(next), w * 0.9f, h * 0.82f, paint);
        }
        paint.setTextSize(h * 0.12f);
        canvas.drawText(Vakit.getByIndex(next).getString(), w * 0.9f, h * 0.95f, paint);


        paint.setColor(Color.WHITE);
        paint.setTextSize(h * 0.25f);
        paint.setTextAlign(Align.CENTER);
        paint.setFakeBoldText(true);
        canvas.drawText(times.getLeft(next, false), w * 0.5f, h * 0.9f, paint);
        paint.setFakeBoldText(false);
        try {
            appWidgetManager.updateAppWidget(widgetId, remoteViews);
        } catch (RuntimeException e) {
            Crashlytics.logException(e);
        }
    }

    @Override
    public void onEnabled(Context context) {
        ComponentName thisWidget = new ComponentName(context, WidgetProviderClock.class);
        AppWidgetManager manager = AppWidgetManager.getInstance(context);
        onUpdate(context, manager, manager.getAppWidgetIds(thisWidget));

    }

    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {

        for (int widgetId : appWidgetIds) {
            updateAppWidget(context, appWidgetManager, widgetId);
        }

    }

    @Override
    public void onDisabled(Context context) {

    }

    @Override
    public void onAppWidgetOptionsChanged(Context context, AppWidgetManager appWidgetManager, int appWidgetId, Bundle newOptions) {
        int w = newOptions.getInt(AppWidgetManager.OPTION_APPWIDGET_MAX_WIDTH);
        int h = newOptions.getInt(AppWidgetManager.OPTION_APPWIDGET_MAX_HEIGHT);

        if (w * h != 0) {
            SharedPreferences widgets = context.getSharedPreferences("widgets", 0);
            SharedPreferences.Editor edit = widgets.edit();
            edit.putInt(appWidgetId + "_width", w);
            edit.putInt(appWidgetId + "_height", h);
            edit.apply();
        }

        updateAppWidget(context, appWidgetManager, appWidgetId);
    }
}

File: features/widget/src/main/java/com/metinkale/prayerapp/vakit/WidgetProviderClock.kt
/*
 * Copyright (c) 2013-2023 Metin Kale
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.metinkale.prayerapp.vakit

import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.ComponentName
import android.content.ContentUris
import android.content.Context
import android.content.Intent
import android.icu.util.Calendar
import android.os.Bundle
import android.os.SystemClock
import android.provider.AlarmClock
import android.provider.CalendarContract
import android.text.Spannable
import android.text.SpannableString
import android.text.style.RelativeSizeSpan
import android.text.style.SuperscriptSpan
import android.util.TypedValue
import android.widget.RemoteViews
import com.metinkale.prayer.Module
import com.metinkale.prayer.Preferences
import com.metinkale.prayer.date.HijriDate
import com.metinkale.prayer.times.fragments.TimesFragment
import com.metinkale.prayer.times.times.Times
import com.metinkale.prayer.times.times.Vakit
import com.metinkale.prayer.times.times.getCurrentTime
import com.metinkale.prayer.times.times.getNextTime
import com.metinkale.prayer.times.times.getTime
import com.metinkale.prayer.times.times.isKerahat
import com.metinkale.prayer.utils.LocaleUtils
import com.metinkale.prayer.utils.UUID
import com.metinkale.prayer.utils.Utils
import com.metinkale.prayer.widgets.R
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.ZoneId

class WidgetProviderClock : AppWidgetProvider() {
    override fun onEnabled(context: Context) {
        val thisWidget = ComponentName(context, WidgetProviderClock::class.java)
        val manager = AppWidgetManager.getInstance(context)
        onUpdate(context, manager, manager.getAppWidgetIds(thisWidget))
    }

    override fun onUpdate(
        context: Context,
        appWidgetManager: AppWidgetManager,
        appWidgetIds: IntArray
    ) {
        for (widgetId in appWidgetIds) {
            updateAppWidget(context, appWidgetManager, widgetId)
        }
    }

    override fun onDisabled(context: Context) {}
    override fun onAppWidgetOptionsChanged(
        context: Context,
        appWidgetManager: AppWidgetManager,
        appWidgetId: Int,
        newOptions: Bundle
    ) {
        updateAppWidget(context, appWidgetManager, appWidgetId)
    }

    companion object {
        fun updateAppWidget(context: Context, appWidgetManager: AppWidgetManager, widgetId: Int) {
            LocaleUtils.init(context)
            val times: Times? = WidgetUtils.getTimes(widgetId)
            if (times == null) {
                WidgetUtils.showNoCityWidget(context, appWidgetManager, widgetId)
                return
            }
            val size = WidgetUtils.getSize(context, appWidgetManager, widgetId, 500f / 200f)
            val width = size.width
            val height = size.height
            if (width <= 0 || height <= 0) return
            val remoteViews = RemoteViews(context.packageName, R.layout.widget_4x2_clock)
            val pendingIntent = TimesFragment.getPendingIntent(times)
            val pendingIntentClock = PendingIntent.getActivity(
                context,
                UUID.asInt(),
                Intent(AlarmClock.ACTION_SHOW_ALARMS),
                PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
            )

            val intent = Module.CALENDAR.buildIntent(context)
            intent.addCategory(Intent.CATEGORY_BROWSABLE)
            val pendingHijri = PendingIntent.getActivity(
                context,
                UUID.asInt(),
                intent,
                PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
            )
            val startMillis = System.currentTimeMillis()
            val builder = CalendarContract.CONTENT_URI.buildUpon()
            builder.appendPath("time")
            ContentUris.appendId(builder, startMillis)
            val calendarIntent = Intent(Intent.ACTION_VIEW).setData(builder.build())
            val pendingIntentCalendar = PendingIntent.getActivity(
                context,
                UUID.asInt(),
                calendarIntent,
                PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
            )
            remoteViews.setOnClickPendingIntent(R.id.time, pendingIntentClock)
            remoteViews.setOnClickPendingIntent(R.id.greg, pendingIntentCalendar)
            remoteViews.setOnClickPendingIntent(R.id.hicri, pendingHijri)
            remoteViews.setOnClickPendingIntent(R.id.lastTime, pendingIntent)
            remoteViews.setOnClickPendingIntent(R.id.lastText, pendingIntent)
            remoteViews.setOnClickPendingIntent(R.id.nextTime, pendingIntent)
            remoteViews.setOnClickPendingIntent(R.id.nextText, pendingIntent)
            remoteViews.setOnClickPendingIntent(R.id.countdown, pendingIntent)
            remoteViews.setViewPadding(R.id.padder, width, height, 0, 0)
            if (Preferences.DIGITS == "normal") {
                if (Preferences.CLOCK_12H) {
                    val cal = Calendar.getInstance()
                    var ampm = "AM"
                    if (cal[Calendar.AM_PM] == Calendar.PM) {
                        ampm = "PM"
                    }
                    val span: Spannable = SpannableString("hh:mm'$ampm'")
                    span.setSpan(SuperscriptSpan(), 5, 8, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
                    span.setSpan(RelativeSizeSpan(0.3f), 5, 8, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
                    remoteViews.setCharSequence(R.id.time, "setFormat12Hour", span)
                    remoteViews.setCharSequence(R.id.time, "setFormat24Hour", span)
                } else {
                    remoteViews.setCharSequence(R.id.time, "setFormat12Hour", "HH:mm")
                    remoteViews.setCharSequence(R.id.time, "setFormat24Hour", "HH:mm")
                }
            } else {
                val txt = LocaleUtils.formatTimeForHTML(LocalTime.now())
                remoteViews.setCharSequence(R.id.time, "setFormat12Hour", txt)
                remoteViews.setCharSequence(R.id.time, "setFormat24Hour", txt)
            }
            val next = times.getNextTime()
            val rtl = Utils.isRTL(context)
            remoteViews.setTextViewText(
                if (!rtl) R.id.lastText else R.id.nextText,
                Vakit.getByIndex(next - 1).string
            )
            remoteViews.setTextViewText(
                if (!rtl) R.id.nextText else R.id.lastText,
                Vakit.getByIndex(next).string
            )
            remoteViews.setTextViewText(
                if (!rtl) R.id.lastTime else R.id.nextTime, LocaleUtils.formatTimeForHTML(
                    times.getTime(
                        LocalDate.now(), next - 1
                    ).toLocalTime()
                )
            )
            remoteViews.setTextViewText(
                if (!rtl) R.id.nextTime else R.id.lastTime, LocaleUtils.formatTimeForHTML(
                    times.getTime(
                        LocalDate.now(), next
                    ).toLocalTime()
                )
            )
            remoteViews.setTextViewText(
                if (!rtl) R.id.greg else R.id.hicri, LocaleUtils.formatDate(
                    LocalDate.now()
                )
            )
            remoteViews.setTextViewText(
                if (!rtl) R.id.hicri else R.id.greg,
                LocaleUtils.formatDate(HijriDate.now())
            )
            remoteViews.setTextViewTextSize(R.id.time, TypedValue.COMPLEX_UNIT_PX, height * 0.6f)
            remoteViews.setTextViewTextSize(R.id.greg, TypedValue.COMPLEX_UNIT_PX, height / 9f)
            remoteViews.setTextViewTextSize(R.id.hicri, TypedValue.COMPLEX_UNIT_PX, height / 9f)
            remoteViews.setTextViewTextSize(R.id.lastTime, TypedValue.COMPLEX_UNIT_PX, height / 6f)
            remoteViews.setTextViewTextSize(R.id.nextTime, TypedValue.COMPLEX_UNIT_PX, height / 6f)
            remoteViews.setTextViewTextSize(R.id.lastText, TypedValue.COMPLEX_UNIT_PX, height / 9f)
            remoteViews.setTextViewTextSize(R.id.nextText, TypedValue.COMPLEX_UNIT_PX, height / 9f)
            remoteViews.setTextViewTextSize(R.id.countdown, TypedValue.COMPLEX_UNIT_PX, height / 5f)
            if (Preferences.COUNTDOWN_TYPE == Preferences.COUNTDOWN_TYPE_SHOW_SECONDS) remoteViews
                .setChronometer(
                    R.id.countdown,
                    times.getTime(LocalDate.now(), next)
                        .atZone(ZoneId.systemDefault()).toInstant()
                        .toEpochMilli() - (System.currentTimeMillis() - SystemClock.elapsedRealtime()),
                    null,
                    true
                ) else {
                val txt = LocaleUtils.formatPeriod(
                    LocalDateTime.now(),
                    times.getTime(LocalDate.now(), next),
                    false
                )
                remoteViews.setString(R.id.countdown, "setFormat", txt)
                remoteViews.setChronometer(R.id.countdown, 0, txt, false)
            }
            remoteViews.setViewPadding(R.id.progresscontainer, width / 10, 0, width / 10, 0)
            remoteViews.setViewPadding(R.id.time, 0, -height / 6, 0, -height / 7)
            remoteViews.setViewPadding(R.id.greg, width / 10, 0, 0, 0)
            remoteViews.setViewPadding(R.id.hicri, 0, 0, width / 10, 0)
            remoteViews.setViewPadding(R.id.lastTime, width / 10, 0, width / 10, -width / 60)
            remoteViews.setViewPadding(R.id.lastText, width / 10, 0, width / 10, 0)
            remoteViews.setViewPadding(R.id.nextTime, width / 10, 0, width / 10, -width / 60)
            remoteViews.setViewPadding(R.id.nextText, width / 10, 0, width / 10, 0)
            val w = width * 10 / 8
            remoteViews.setInt(
                if (!rtl) R.id.progressBg else R.id.progress,
                "setBackgroundColor",
                -0x1
            )
            remoteViews.setInt(
                if (!rtl) R.id.progress else R.id.progressBg,
                "setBackgroundColor",
                if (times.isKerahat()) -0x40c0a5 else Theme.Light.strokecolor
            )
            var passedPart = getPassedPart(times)
            if (rtl) passedPart = 1 - passedPart
            remoteViews.setViewPadding(R.id.progress, (w * passedPart).toInt(), width / 75, 0, 0)
            remoteViews.setViewPadding(
                R.id.progressBg,
                (w * (1 - passedPart)).toInt(),
                width / 75,
                0,
                0
            )
            appWidgetManager.updateAppWidget(widgetId, remoteViews)
        }

        private fun getPassedPart(times: Times): Float {
            val current = times.getCurrentTime()
            val now = LocalDateTime.now().atZone(ZoneId.systemDefault()).toInstant().toEpochMilli()
            val prev =
                times.getTime(LocalDate.now(), current).atZone(ZoneId.systemDefault()).toInstant()
                    .toEpochMilli()
            val next =
                times.getTime(LocalDate.now(), current + 1).atZone(ZoneId.systemDefault())
                    .toInstant()
                    .toEpochMilli()
            return (now - prev) / (next - prev).toFloat()
        }

    }
}